### PR TITLE
[Feature] Quantum dropout

### PR DESCRIPTION
**Pull request by callumfduffy**
This is a work in progress example for quantum dropout.

Closes https://github.com/pasqal-io/pyqtorch/issues/163

TODOs:

- [x] address all comments @callumfduffy 
- [x] move contents of example.py into `docs/dropout.md` @callumfduffy 
- [x] final review by @gvelikova

### CODE REVIEW

**Comment by dominikandreasseitz** [comment]
hi @callumfduffy, great work! can you describe the batching issue youre facing?

**Comment by gvelikova** [comment]
@dominikandreasseitz I think the batching issue was resolved and the description is just a tad stale.

File: pyqtorch/circuit.py [hunk 1]
```diff
@@ -122,6 +122,84 @@ def flatten(self) -> ModuleList:
...
+        if self.training:
+            state = self.dropout_fn(state, values)
+        else:
+            for op in self.operations:
+                state = op(state, values)
+        return state
+
+    def rotational_dropout(
+        self, state: State = None, values: dict[str, Tensor] | ParameterDict = {}
+    ) -> State:
```
Comment (dominikandreasseitz) [inline pyqtorch/circuit.py:149]: can we add some docstrings short description of what the method is doing?

File: pyqtorch/utils.py [hunk 2]
```diff
@@ -64,6 +64,22 @@ class DiffMode(StrEnum):
     ADJOINT = "adjoint"
 
 
+class DropoutMode(StrEnum):
+    """
+    Which Dropout mode to use.
+
+    Options: none          - No dropout is used.
```
Comment (dominikandreasseitz) [inline pyqtorch/utils.py:71]: you can remove "none"

File: pyqtorch/circuit.py [hunk 3]
```diff
@@ -122,6 +122,84 @@ def flatten(self) -> ModuleList:
...
+        dropout_mode: DropoutMode = DropoutMode.ROTATIONAL,
+        dropout_prob: float = 1.0,
+    ):
+        super().__init__(n_qubits, operations)
+        self.dropout_mode = dropout_mode
+        self.dropout_prob = dropout_prob
+
+        self.dropout_fn = getattr(self, dropout_mode)
+
+    def forward(self, state: State, values: dict[str, Tensor] | ParameterDict = {}) -> State:
```
Comment (dominikandreasseitz) [inline pyqtorch/circuit.py:139]: we dont need the if...elif. we are always using dropout now right?

**Review by dominikandreasseitz — CHANGES_REQUESTED** [review]
hey @callumfduffy nice work, i put some comments. can you also add some documentation under `docs/index.md` (just add another section) with some more details like links to papers, short description of the used algorithms etc.

File: docs/index.md [hunk 4]
```diff
@@ -379,6 +379,367 @@ def fig_to_html(fig: Figure) -> str:  # markdown-exec: hide
 print(fig_to_html(plt.gcf())) # markdown-exec: hide
 ```
 
+## Fitting a noisy sinusoid with quantum dropout
+
```
Comment (dominikandreasseitz) [inline docs/index.md:383]: hey @callumfduffy can you put this into `docs/dropout.md`  (rebase on main if you havent)

File: pyqtorch/circuit.py (hunk 3 above)
Comment (callumfduffy) [reply pyqtorch/circuit.py]: kept in case you want to use the circuit at test time and do not want dropout present

File: docs/dropout.md [hunk 5]
```diff
@@ -1 +1,361 @@
...
+
+import pyqtorch as pyq
+from pyqtorch.circuit import DropoutQuantumCircuit
+from pyqtorch.parametric import Parametric
+
+manual_seed(12345)
+np.random.seed(12345)
+
+
+class MinMaxScaler:
```
Comment (gvelikova) [inline docs/dropout.md:25]: Ok, for all of the  functions and classes here I would suggest to add docstrings and comments (simpler functions may have just a sentence of what they are), but this way their functionalities will be clearer to the user.

Comment (gvelikova) [reply docs/dropout.md:25]: Also, for this documentation treat it as a proper tutorial, which implies that you need to add more context on each of the steps taken. Check the tutorials in qadence.

File: pyqtorch/circuit.py [hunk 6]
```diff
@@ -215,6 +215,112 @@ def sample(probs: Tensor) -> Counter:
             return list(map(lambda p: sample(p), probs))
 
 
+class DropoutQuantumCircuit(QuantumCircuit):
+    def __init__(
```
Comment (gvelikova) [inline pyqtorch/circuit.py:219]: Again as stated above, add docstrings.

File: pyqtorch/circuit.py [hunk 7]
```diff
@@ -215,6 +215,112 @@ def sample(probs: Tensor) -> Counter:
...
+        self,
+        n_qubits: int,
+        operations: list[Module],
+        dropout_mode: DropoutMode = DropoutMode.ROTATIONAL,  # type: ignore
```
Comment (gvelikova) [inline pyqtorch/circuit.py:223]: Ideally, after looking into the StringEnum you would be able to unmute mypy type checks.

File: pyqtorch/circuit.py [hunk 8]
```diff
@@ -215,6 +215,112 @@ def sample(probs: Tensor) -> Counter:
...
+                    keep = 1 - bernoulli(tensor(self.dropout_prob))  # type: ignore
+                    if keep:
+                        state = op(state, values)
+
+                else:
+                    state = op(state, values)
+            else:
+                state = op(state, values)
+
+        return state
```
Comment (gvelikova) [inline pyqtorch/circuit.py:266]:
``` # mild suggestion if it works
for op in self.operations:
    if hasattr(op, "param_name") and values[op.param_name].requires_grad and 1 - bernoulli(tensor(self.dropout_prob)):
        return state
    else:
        return op(state, values) # new state
```

Comment (gvelikova) [reply pyqtorch/circuit.py:266]: Potentially, if useful this can be applied to the code of the other types of dropout.

File: pyqtorch/utils.py [hunk 9]
```diff
@@ -87,6 +87,21 @@ class DiffMode(StrEnum):
     ADJOINT = "adjoint"
 
 
+class DropoutMode(StrEnum):
+    """
+    Which Dropout mode to use.
```
Comment (gvelikova) [inline pyqtorch/utils.py:92]: Maybe expand on this.

**Review by dominikandreasseitz — CHANGES_REQUESTED** [review]
Hi @callumfduffy. Thanks a lot! Some remarks:

- [x] Regarding your tutorial, Please have a look at https://github.com/pasqal-io/pyqtorch/blob/main/docs/fitting_a_function.md, i use a `hea` fixture there + maybe you can steal some of the logic to create your circuit from the circuit there too
- [x] In the same tutorial, you will find some lines at the bottom which display your figures, please just copy them into your tutorial to display your results
- [x] Please remove all the type ignores in your code if possible
- [x] Please reduce the number of epochs significantly so the docs build faster. Also raise the batch_size to 500 to you get a better fit

**Comment by dominikandreasseitz** [comment]
hi @callumfduffy , @Roland-djee will take over the review here, thanks!

File: docs/dropout.md [hunk 10]
```diff
@@ -1 +1,402 @@
...
+# the two feature maps we will be using
+def fm1(n_qubits):
+    return [pyq.RY(i, "x1") for i in range(n_qubits)]
+
+
+def fm2(n_qubits):
+    return [pyq.RZ(i, "x2") for i in range(n_qubits)]
+
+# The class which constructs QNNs
+class QuantumModelBase(torch.nn.Module):
```
Comment (Roland-djee) [inline docs/dropout.md:176]: I'm a bit surprised there is no datastructure you could use here instead of reconstructing one.

File: pyqtorch/circuit.py [hunk 11]
```diff
@@ -215,6 +215,114 @@ def sample(probs: Tensor) -> Counter:
...
+        for op in self.operations:
+            if (
+                (hasattr(op, "param_name"))
+                and (values[op.param_name].requires_grad)
+                and not (int(1 - bernoulli(tensor(self.dropout_prob))))
+            ):
+                continue
+
+            else:
+                state = op(state, values)
```
Comment (Roland-djee) [inline pyqtorch/circuit.py:269]: I think I would simplify it by doing something like:

```suggestion
            has_param_name = hasattr(op, "param_name")
            requires_grad = values[op.param_name].requires_grad
            not_<name> = not (int(1 - bernoulli(tensor(self.dropout_prob)))
            if not (has_param_name and requires_grad not_<name>):
                state = op(state, values)
```

File: pyqtorch/circuit.py [hunk 12]
```diff
@@ -215,6 +215,114 @@ def sample(probs: Tensor) -> Counter:
...
+        Returns:
+            State: pure state vector
+        """
+        for op in self.operations:
+            if not (hasattr(op, "param_name")) and not (
+                int(1 - bernoulli(tensor(self.dropout_prob)))
+            ):
+                continue
+            else:
+                state = op(state, values)
```
Comment (Roland-djee) [inline pyqtorch/circuit.py:290]: Same. I would define a bunch of predicates and remove the useless `continue` clause.

File: pyqtorch/circuit.py [hunk 13]
```diff
@@ -215,6 +215,114 @@ def sample(probs: Tensor) -> Counter:
             return list(map(lambda p: sample(p), probs))
 
 
+class DropoutQuantumCircuit(QuantumCircuit):
```
Comment (Roland-djee) [inline pyqtorch/circuit.py:218]: Is there a need for another class or could this be integrated in `QuantumCircuit` ?

**Review by Roland-djee — COMMENTED** [review]
Hey thanks @callumfduffy. Comments from my side.

File: pyqtorch/circuit.py (hunk 11 above)
Comment (callumfduffy) [reply pyqtorch/circuit.py]: unfortunately I don't believe we can predicates beforehand since "has_param_name = hasattr(op, "param_name")" will return an error for gates which do not have the param_name attribute such as the CNOT gate. The if statement avoids this issue since the statements are evaluated sequentially. However, I can re-write this so we no longer have the continue statement and just one if statement.

File: pyqtorch/circuit.py (hunk 13 above)
Comment (callumfduffy) [reply pyqtorch/circuit.py:218]: This was a suggestion from @dominikandreasseitz, so that the QuantumCircuit class remained not overly complex.

Comment (jpmoutinho) [reply pyqtorch/circuit.py:218]: I agree with designing this on a separate class. All methods here are too specific to put in the QuantumCircuit.

File: docs/dropout.md (hunk 10 above)
Comment (jpmoutinho) [reply docs/dropout.md:176]: No, there isn't. So far there wasn't really a need for one. 

It seems to me like the usage here is just to follow the typical torch neural network construction, which in a way is a more "torchy" use of PyQ than what Qadence actually does. 

For now, if it's just for convenience of the tutorial that's fine. We can then see if a general version of this deserves a place in PyQ.

File: docs/dropout.md [hunk 14]
```diff
@@ -1 +1,402 @@
...
+
+        self.embedding1 = fm1(n_qubits=n_qubits)
+        self.embedding2 = fm2(n_qubits=n_qubits)
+
+        self.params = torch.nn.ParameterDict()
+        operations = self.build_operations()
+        self.circuit = self.build_circuit(operations)
+        self.observable = pyq.QuantumCircuit(n_qubits, [pyq.Z(i) for i in range(n_qubits)]).to(
+            device=device, dtype=torch.complex64
+        )
```
Comment (jpmoutinho) [inline docs/dropout.md:191]: Please use the `Observable` class to build observables (it's recent).

File: tests/test_dropout.py [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        dropout_prob=1.0,
+    )
+    obs = pyq.QuantumCircuit(n_qubits=n_qubits, operations=[pyq.Z(1)])
+
+    exp_circ = pyq.expectation(circuit=circ, state=state, values=values, observable=obs)
+    exp_dropout_circ = pyq.expectation(
+        circuit=dropout_circ, state=state, values=values, observable=obs
+    )
+
+    assert torch.allclose(exp_circ, exp_dropout_circ, atol=ATOL)
```
Comment (jpmoutinho) [inline tests/test_dropout.py:49]: It is not clear to me what the logic of this test is. If one circuit does dropout and the other doesn't, shouldn't we expect the expectation value to be different?

If we are testing for equality in the expectation value, is it that the dropout configuration is set to never dropout? If yes, than I understand that is a good sanity check to include, but I feel like something more should be tested regarding the actual effect of the dropout methods.

But maybe I am missing something so feel free to correct me 😅

**Review by jpmoutinho — COMMENTED** [review]
Just a quick look from my side, thanks for the work @callumfduffy :). Overall the code looks very nice and clean.

File: tests/test_dropout.py (hunk 15 above)
Comment (callumfduffy) [reply tests/test_dropout.py:49]: Sorry for the confusion I've added a docstring to the test to clear up its purpose. The tests set the dropout probability to 1 such that every gate the specific dropout method could effect is dropped, this provides a deterministic result which we can manually check against.

Comment (jpmoutinho) [reply tests/test_dropout.py:49]: Ah got it, I see now that you are testing, thanks.

**Review by gvelikova — APPROVED** [review]
